docs/architecture/adr-036-blockchain-reactor-refactor.md

```diff
@@ -0,0 +1,438 @@
+}
+
+type ResponseTimeout struct {
+	PeerID ID
+	Height int64
+}
+
+type TerminationTimeout struct {
+	Height int64
+}
+
+type Message int
+
+const (
+	MessageUnknown Message = iota
+	MessageBlockRequest
+	MessageBlockResponse
+	MessageStatusReport
+)
+
+type BlockRequestMessage struct {
+	Height int64
+	PeerID ID
+}
+
+type BlockResponseMessage struct {
+	Height        int64
+	Block         Block
+	Commit        Commit
+	PeerID        ID
+	CurrentHeight int64
+}
+
+type StatusReportMessage struct {
+	Height int64
+}
+
+type TimeoutTrigger int
+
+const (
+	TimeoutUnknown TimeoutTrigger = iota
+	TimeoutResponseTrigger
+	TimeoutTerminationTrigger
+)
+```
+
+The Controller state machine can be in two modes (states): `FastSyncMode` when
+it is trying to catch up with the network by downloading committed blocks,
+and `ConsensusMode` in which it executes Tendermint consensus protocol. We
+assume that "fast sync" mode terminates once the Controller switch to
+`ConsensusMode`.  
+
+``` go
+type Mode int
+
+const (
+	ModeUnknown Mode = iota
+	ModeFastSync
+	ModeConsensus
+)
+
+type ControllerState struct {
+	Height             int64            // the first block that is not committed
+	Mode               Mode             // mode of operation of the state machine
+	PeerMap            map[ID]PeerStats // map of peer IDs to peer statistics
+	MaxRequestPending  int64            // maximum height of the pending requests
+	FailedRequests     []int64          // list of failed block requests
+	PendingRequestsNum int              // number of pending requests
+	Store              []BlockInfo      // contains list of downloaded blocks
+	Executor           BlockExecutor    // store, verify and executes blocks
+}
+
+type PeerStats struct {
+	Height         int64
+	PendingRequest int64 // it can be list in case there are multiple outstanding requests per peer
+}
+
+type BlockInfo struct {
+	Block  Block
+	Commit Commit
+	PeerID ID // a peer from which we downloading the corresponding Block and Commit
+}
+```
+
+The `Controller` is initialized by providing initial height (`startHeight`) from which it will start downloading blocks from peers.
+
+``` go
+func NewControllerState(startHeight int64, executor BlockExecutor) ControllerState {
+	state = ControllerState {}
+    state.Height = startHeight
+	state.Mode = ModeFastSync
+	state.MaxRequestPending = startHeight - 1
+    state.PendingRequestsNum = 0
+    state.Executor = executor
+    initialize state.PeerMap, state.FailedRequests and state.Store to empty data structures
+    return state
+}
+
+func handleEvent(state ControllerState, event Event) (ControllerState, Message, TimeoutTrigger, Error) {
+	msg = nil
+	timeout = nil
+	error = nil
+
+	switch state.Mode {
+	case ModeConsensus:
+		switch event := event.(type) {
+		case EventBlockRequest:
+			msg = createBlockResponseMessage(state, event)
+			return state, msg, timeout, error
+		default:
+			error = "Only respond to BlockRequests while in ModeConsensus!"
```

Comment:
```suggestion
			error = "While in ModeConsensus only respond to Block Requests"
```

What about the status requests? Where do we reply to those?

Reply:
It is part of the Heartbeat mechanism that is part of the Executor logic. I assumed that Executor send periodically to each peer StatusRequest message and expects StatusReport to arrive within response timeout.

Reply:
What's in a name... "peer"?  :)
A node that had previously answered to the status request broadcast? Or the peers the switch knows about (here we don't know the peer fast sync status)? 
If the former, will Executor keep track of peers in order to know who didn't answer?

Reply:
A peer is a node with which we have open connection (at the switch level). Executor sends status requests to all peers and need to manage timeouts. Therefore, we probably need to have peer lists at the Executor level also to able to determine who didn't responded on time.